File: ios/Classes/RxEventChannelPlugin.swift
import Flutter
import UIKit

// 모든 isolate/engine 간에 이벤트를 브로드캐스트하는 싱글톤 클래스
class InterIsolateBroadcaster {
  static let shared = InterIsolateBroadcaster()
  
  // 스레드 안전을 위한 직렬 큐
  private let queue = DispatchQueue(label: "com.joel.rx_event_channel.broadcaster")
  
  // 등록된 이벤트 싱크 모음
  private var sinks = NSMutableSet()
  
  private init() {}
  
  // 새 이벤트 싱크 등록
  func addSink(_ sink: @escaping FlutterEventSink) {
    queue.sync {
      sinks.add(sink)
    }
  }
  
  // 모든 등록된 싱크에 이벤트 브로드캐스트
  func broadcastEvent(_ event: [String: Any]) {
    queue.sync {
      // Swift에서는 NSMutableSet을 반복하면서 요소를 제거할 수 없으므로
      // 별도의 배열에 모든 싱크를 복사한 후 작업
      let allSinks = sinks.allObjects
      
      for case let sink as FlutterEventSink in allSinks {
        // 메인 큐에서 이벤트 전송
        DispatchQueue.main.async {
          sink(event)
        }
      }
    }
  }
}

public class RxEventChannelPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = RxEventChannelPlugin()
    
    // 이벤트 발생용 메서드 채널
    let emitChannel = FlutterMethodChannel(name: "inter_isolate_event/emit", binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(instance, channel: emitChannel)
    
    // 브로드캐스트 이벤트 채널
    let eventChannel = FlutterEventChannel(name: "inter_isolate_event/broadcast", binaryMessenger: registrar.messenger())
    eventChannel.setStreamHandler(instance)
  }
  
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if call.method == "emitEvent", let args = call.arguments as? [String: Any] {
      guard let eventType = args["eventType"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "eventType is required", details: nil))
        return
      }
      
      let payload = args["payload"]
      let event: [String: Any] = [
        "eventType": eventType,
        "payload": payload as Any
      ]
      
      InterIsolateBroadcaster.shared.broadcastEvent(event)
      result(nil)
    } else {
      result(FlutterMethodNotImplemented)
    }
  }
  
  // MARK: - FlutterStreamHandler 메서드
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    InterIsolateBroadcaster.shared.addSink(events)
    return nil
  }
  
  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    // iOS에서는 싱크가 무효화되면 자동으로 가비지 컬렉션됨
    return nil
  }
}

File: android/src/main/kotlin/com/joel/rx_event_channel/rx_event_channel/RxEventChannelPlugin.kt
package com.joel.rx_event_channel.rx_event_channel

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.CopyOnWriteArraySet

/** RxEventChannelPlugin */
class RxEventChannelPlugin: FlutterPlugin, EventChannel.StreamHandler {
  // 이벤트 채널 (브로드캐스트용)
  private lateinit var eventChannel: EventChannel
  // 이벤트 발생용 메서드 채널
  private lateinit var emitChannel: MethodChannel
  
  // 싱글톤 인스턴스 관리
  companion object {
    private val instance = InterIsolateBroadcaster()
    
    fun getInstance(): InterIsolateBroadcaster {
      return instance
    }
  }

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    // 이벤트 발생용 메서드 채널
    emitChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "inter_isolate_event/emit")
    emitChannel.setMethodCallHandler { call, result ->
      if (call.method == "emitEvent") {
        val eventType = call.argument<String>("eventType")
        val payload = call.argument<Any>("payload")
        
        if (eventType != null) {
          val event = mapOf(
            "eventType" to eventType,
            "payload" to payload
          )
          getInstance().broadcastEvent(event)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "eventType is required", null)
        }
      } else {
        result.notImplemented()
      }
    }
    
    // 브로드캐스트 이벤트 채널
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "inter_isolate_event/broadcast")
    eventChannel.setStreamHandler(this)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    emitChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }
  
  // EventChannel.StreamHandler 구현
  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    events?.let { 
      getInstance().addSink(it) 
    }
  }
  
  override fun onCancel(arguments: Any?) {
    // 취소 로직 간소화 - Flutter 측에서 구독 취소 시 자동으로 처리됨
    // 실제 싱크 제거는 broadcastEvent에서 예외 발생 시 처리
  }
}

/**
 * 모든 isolate/engine 간에 이벤트를 브로드캐스트하는 싱글톤 클래스
 */
class InterIsolateBroadcaster {
  // Thread-safe한 이벤트 싱크 컬렉션
  private val sinks = CopyOnWriteArraySet<EventChannel.EventSink>()
  
  /**
   * 새 이벤트 싱크 등록
   */
  fun addSink(sink: EventChannel.EventSink) {
    sinks.add(sink)
  }
  
  /**
   * 모든 등록된 싱크에 이벤트 브로드캐스트
   */
  fun broadcastEvent(event: Map<String, Any?>) {
    // 모든 싱크에 이벤트 전송
    for (sink in sinks) {
      try {
        sink.success(event)
      } catch (e: Exception) {
        // 실패한 싱크는 즉시 제거 (CopyOnWriteArraySet은 반복 중 제거 가능)
        sinks.remove(sink)
      }
    }
  }
}
